src/Instrumentation/Psr16/src/Psr16Instrumentation.php

```diff
@@ -0,0 +1,180 @@
+                $span = self::makeSpanBuilder($instrumentation, $function, $class, $filename, $lineno)
+                    ->startSpan();
+
+                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
+            },
+            post: static function (CacheInterface $cacheItem, array $params, $return, ?Throwable $exception) {
+                self::end($exception);
+            }
+        );
+
+        hook(
+            CacheInterface::class,
+            'setMultiple',
+            pre: static function (CacheInterface $cacheItem, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                $span = self::makeSpanBuilder($instrumentation, $function, $class, $filename, $lineno)
+                    ->startSpan();
+
+                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
+            },
+            post: static function (CacheInterface $cacheItem, array $params, $return, ?Throwable $exception) {
+                self::end($exception);
+            }
+        );
+
+        hook(
+            CacheInterface::class,
+            'deleteMultiple',
+            pre: static function (CacheInterface $cacheItem, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                $span = self::makeSpanBuilder($instrumentation, $function, $class, $filename, $lineno)
+                    ->startSpan();
+
+                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
+            },
+            post: static function (CacheInterface $cacheItem, array $params, $return, ?Throwable $exception) {
+                self::end($exception);
+            }
+        );
+
+        hook(
+            CacheInterface::class,
+            'has',
+            pre: static function (CacheInterface $cacheItem, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                $span = self::makeSpanBuilder($instrumentation, $function, $class, $filename, $lineno)
+                    ->startSpan();
+
+                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
+            },
+            post: static function (CacheInterface $cacheItem, array $params, $return, ?Throwable $exception) {
+                self::end($exception);
+            }
+        );
+    }
+
+    private static function makeSpanBuilder(
+        CachedInstrumentation $instrumentation,
+        string $function,
+        string $class,
+        ?string $filename,
+        ?int $lineno
+    ): SpanBuilderInterface
+    {
+        return $instrumentation->tracer()
+            ->spanBuilder(sprintf('%s::%s', $class, $function))
+            ->setSpanKind(SpanKind::KIND_CLIENT)
```

Comment:
```suggestion
            ->setSpanKind(SpanKind::KIND_INTERNAL)
```

A cache interaction usually involves outgoing calls to redis, memcached or another implementation. If these had instrumentation active, we'd create CLIENT spans with CLIENT children, which is semantically incorrect. I propose we either detect if other instrumentation is active and make the kind contingent on that, or make these spans internal per se

Reply:
When I was making the changes discussed, I ended up having some questions:
- Would making all spans internal be violating the semantic convention of maintaining attributes (`db.operation`, `db.system`)?
- In the image below I used [this in-memory cache file](https://github.com/open-telemetry/opentelemetry-php-contrib/pull/244/files#diff-ed82a484ae6e7f8cbae2e02d761ebe71f99ba9791404318f23b52c1bf3f0b2bbR103) to perform some tests, so I had a question that is related to your comment when continuing to create spans as a Client, in cases such as the deleteMultiple operation that can generate child spans, these must be assigned like Internal ? Could you get an example of how to detect an active Client interval?

![traces](https://github.com/open-telemetry/opentelemetry-php-contrib/assets/54692047/919d4a9e-37cb-472e-a9d2-63fbb116e245)

Reply:
@cedricziel could you take a look at these questions?

Reply:
I think the part of the spec Cedric is referring to is https://opentelemetry.io/docs/concepts/signals/traces/#span-kind
> "... the parent of a server span is often a remote client span, and the child of a client span is usually a server span ..."

It's not a MUST though, but let's assume that it is semantically incorrect to have client->client. So, let's make the spans internal, and not use DB attributes. How about using `cache.xxx` as the attribute names? Then, if the underlying implementation _is_ a database, it should set `db.xxx` as appropriate.

I think that's good enough for an initial release.

Reply:
I like the idea, I was thinking about at least having some attributes like:
- In the operations `get, set, has, getMultiple, has ...` have an attribute like cache.operation
- In` get, set, getMultiple, setMultiple, deleteMultiple and has` operations have a `cache.key` or `cache.keys` attribute to identify the keys used.

These would be some attributes that I see that can make up the span. What do you think ?

Reply:
I think that's good, do it. I'm leaning towards using `cache.keys` and making it an array of strings for the second part.